File: BehaviorTree/Branch/Decorator/Timer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Timers;

namespace FluentBehaviorTree
{
    public class Timer : Decorator
    {
        private System.Timers.Timer timer;

        /// <summary>
        /// Reset its inner state
        /// </summary>
        protected override void Open()
        {
            base.Open();
            innerResult = Status.RUNNING;
            timerSet = false;
        }

        /// <summary>
        /// ElapsedEventHandler
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void OnTimeout(object sender, System.Timers.ElapsedEventArgs e)
        {
            innerResult = child.Tick();
            timer.Elapsed -= OnTimeout;
            timerSet = false;
        }

        private bool timerSet = false;

        private Status innerResult = Status.RUNNING;

        public Timer(string name, ulong ms)
        {
            this.Name = name;
            timer = new System.Timers.Timer(ms);
        }

        /// <summary>
        /// Wait until ms is reached and run child. Return running during that time
        /// </summary>
        /// <returns></returns>
        protected override Status tick()
        {
            if (timer.Enabled)
            {
                this.Result = Status.RUNNING;
            }
            else if (innerResult != Status.RUNNING)
            {
                this.Result = innerResult;
            }
            else
            {
                if (!timerSet)
                {
                    timer.AutoReset = false;
                    timer.Elapsed += OnTimeout;
                    timerSet = true;
                    timer.Enabled = true;
                }

                this.Result = Status.RUNNING;
            }
            return this.Result;
        }
    }
}

File: BehaviorTreesLib/Timer.cpp
#include "Timer.h"
#include <thread>
#include <chrono>

namespace fluentBehaviorTree
{

	void Timer::onTimeout()
	{
		mInnerResult = mChild->tick();
		timerSet = false;
	}

	void Timer::setTimer()
	{
		std::this_thread::sleep_for(std::chrono::milliseconds(mMS));
		this->onTimeout();
	}

	Timer::Timer(std::string name, size_t ms)
	{
		this->setName(name);
		mMS = ms;
		mInnerResult = EStatus::RUNNING;
		timerSet = false;
	}

	// Reset its inner state
	void Timer::open()
	{
		Node::open();
		mInnerResult = EStatus::RUNNING;
		timerSet = false;
	}

	// Wait until ms is reached and run child. Return running during that time
	EStatus Timer::tickNode()
	{
		if (timerSet)
		{
			this->setResult(EStatus::RUNNING);
		}
		else if (mInnerResult != EStatus::RUNNING)
		{
			this->setResult(mInnerResult);
		}
		else
		{
			if (!timerSet)
			{
				std::thread timer(&Timer::setTimer, this);
				timerSet = true;
				timer.detach();
			}

			this->setResult(EStatus::RUNNING);
		}
		return this->getResult();
	}
}